doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +668,381 @@ Applying theorems
+
+      .. coqtop:: all
+
+         apply Nat.le_trans with (y * x).
+
+   .. _apply_with_binding_hyp:
+   .. example:: Apply a theorem with a binding in a hypothesis
+
+      When applying a theorem in a hypothesis,
+      :tacn:`apply` unifies the hypothesis with one of the premises
+      of the theorem `le_trans : forall n m p, n <= m -> m <= p -> n <= p`.
+      In this case, it unifies with the first premise
+      (`n <= m`) and assigns `x * x` and `y * y` to,
+      respectively, `n` and `m` in the theorem (forward reasoning).
+      The  `with` clause provides the binding for `p`.
+
+      In addition, :tacn:`apply` in a hypothesis isn't as flexible as
+      :tacn:`apply` in the goal: for hypotheses, the unbound variable can be bound
+      by name (as shown) or values for all the variables can be given
+      positionally, i.e. `apply Nat.le_trans with (x * x) (y * y) (y * x) in H.`
+
+      .. coqtop:: reset in
+
+         Require Import PeanoNat.
+
+      .. coqtop:: none
+
+         Goal forall (x y : nat), x * x <= y * y -> x <= y.
+
+      .. coqtop:: out
+
+         intros x y H.
+
+      .. coqtop:: all
+
+         apply Nat.le_trans with (p := y * x) in H.
+
+   .. _apply_with_iff:
+   .. example:: Applying theorems with `<->`
+
+      .. Note: :n:`/\` and :n:`/\\` don't give the desired output.  A bug.
 
-      .. flag:: Apply With Renaming
+      :n:`A <-> B` is defined as :n:`(A -> B) /\ (B -> A)`.
+      `/\\` represents an inductive type with a single constructor:
+      :n:`Inductive and (C D:Prop) : Prop := conj : C -> D -> D /\ C`.  The premises
+      of :n:`conj` are :n:`C` and :n:`D`.  The tactic tries the constructor premises
+      in left-to-right order and uses the first match.
 
-         When on, this flag causes the names in the :n:`@term`'s type to be renamed for uniqueness.
-         By default no renaming is done.
+      Theorems that use :n:`<->` to state a logical equivalence behave consistently
+      when applied to goals and hypotheses.
 
-         .. deprecated:: 8.15
+      .. coqtop:: reset none
+
+         Goal forall (H1: 1 = 1 <-> 2 = 2) (H: 1 = 1), 1 = 1.
+
+      .. coqtop:: out
 
-            This flag is provided for compatibility with versions prior to 8.15.
+         intros H1 H.
 
-   .. tacv:: apply {+, @term}
+      .. coqtop:: all
+
+         apply H1.
+         apply H1 in H.
 
-      This is a shortcut for :n:`apply @term__1; [.. | ... ; [ .. | apply @term__n] ... ]`,
-      i.e. for the successive applications of :n:`@term`:sub:`i+1` on the last subgoal
-      generated by :n:`apply @term__i` , starting from the application of :n:`@term__1`.
+   .. _example_apply_pattern:
+   .. example:: Using :tacn:`pattern` to control unification in apply
 
-   .. tacv:: eapply @term
-      :name: eapply
+      After the ``pattern``, the ``apply`` uses the special case second-order
+      unification described :ref:`here <apply_with_second_order_unification>`
+      (after "unless").
```

Comment:
Consistently with my remarks above, I'm convinced that the "After the `pattern`" part should be removed. You can just say that this example illustrates the unification behavior described in the referenced paragraph.

Reply:
Good now?